prow/cmd/tide/README.md

```diff
@@ -0,0 +1,81 @@
+# Tide Documentation
+
+Tide merges PR that match a given sets of criteria
+
+## Tide configuration
+
+Extend the primary prow [`config.yaml`] document to include a top-level
+`tide` key that looks like the following:
+
+```yaml
+
+tide:
+  queries:
+  ...
+  branchProtection:
+  ...
+  merge_method:
+  ...
+
+
+presubmits:
+  kubernetes/test-infra:
+  - name: fancy-job-name
+    context: fancy-job-name
+    always_run: true
+    spec:  # podspec that runs job
+```
+
+
+### Merging Options
+
+Tide supports all 3 github merging options:
+
+* squash
+* merge
+* rebase
+
+A merge method can be set for repo or per org.
+
+Example:
+
+```yaml
+tide:
+  ...
+  merge_method:
+    org1: squash
+    org2/repo1: rebase
+    org2/repo2: merge
+```
+
+### Queries Configuration
+
+Queries are using github queries to find PRs to be merged. Multiple queries can be defined for a single repo. Queries support filtering per existing and missing labels. In order to filter PRs that have been approved, use the reviewApprovedRequired.
+
+```yaml
+tide:
+  queries:
+    ...
+    - repos:
+      - org1/repo1
+      - org2/repo2
+      labels:
+      - labelThatMustsExists
+      - OtherLabelThatMustsExist
+      missingLabels:
+      - labelThatShouldNotBePresent
+     # If you want github approval
+     reviewApprovedRequired: true
+```
+
+### Branch Protection Options
+Branch Protection options are use to enforce github branch protection.
+A PR will be merged when all required checks are passing, meaning we will skip optional contexts.
```

Comment:
So if I understand this correctly, branch protection configuration stored in github specifies required contexts. Also `optional` in the presubmits specifies required contexts from the prow side. Is tide going to merge based on the union of these contexts?

Reply:
So instead of using the data stored in github (querying github for required check on every PR consumes too many token), we use the branch protection data stored in the config. The branch protection parse the config + the prow jobs settings and all that information goes into the contextRegister